File: FourierStatsMatlab/tsqh_test.m
function output = tsqh_test(x,y,paired,mu)

% tsqh.test: function to calculate a different variants of the T-squared test of Hotelling (1931)
% the inputs must be Nx2 or 2xN matrices of numbers
% for a one-sample test, mu is an optional vector to which the data are compared
% for a two-sample test, x and y are the data from the two conditions
% if paired=TRUE, the matrices (x and y) must be the same size

d = size(x);
if (d(1)<d(2))
    x = x';
end
d = size(y);
if (d(1)<d(2))
    y = y';
end

if (~isempty(mu))
    for n = 1:length(mu)
        x(:,n) = x(:,n) - mu(n);
    end
end

if ~isreal(x)
    x = [real(x); imag(x)]';
end
if (~isempty(y))
    if ~isreal(y)
        y = [real(y); imag(y)]';
    end
end

if (isempty(y))
    method = 'One-sample T-squared test';
end
if (~isempty(y) && paired==1)
    method = 'Paired samples T-squared test';
end
if (~isempty(y) && paired==0)
    method = 'Independent samples T-squared test';
end

if (paired==1)
    if (~isempty(y))
        x = x - y;
    end
end


if (paired==1 || isempty(y))
    % paired or one-sample version of the test

    xbar = mean(x);
    s = size(x);
    m = s(2);
    N = s(1);
    C = cov(x);
    Cinv = inv(C);
    tsq = N * xbar * Cinv * xbar';
    Fratio = tsq .* (N - m)./(m.*(N-1));
    df1 = m;
    df2 = N-m;
    pval = fcdf(Fratio,df1,df2,'upper');
 
else
    % independent samples version of the test
    
    xbarA = mean(x);
    s = size(x);
    m = s(2);
    N1 = s(1);
    
    xbarB = mean(y);
    s = size(y);
    N2 = s(1); 
    
    CA = cov(x);
    CB = cov(y);
    
    meandiff = xbarA - xbarB;
    
    C = ((N1-1)*CA + (N2-1)*CB)/(N1+N2-m);
    Cinv = inv(C);
    
    tsq = ((N1*N2)/(N1+N2)) * meandiff * Cinv * meandiff';
    Fratio = tsq * (N1+N2-m-1)/(m*(N1+N2-2));
    df1 = m;
    df2 = N1+N2-m-1;
    
    pval = fcdf(Fratio,df1,df2,'upper');
    
end

pval = min(pval,1);

output.tsqc = tsq;
output.Fratio = Fratio;
output.df1 = df1;
output.df2 = df2;
output.pval = pval;
output.method = method;

end

File: R/tsqh.test.R
#' tsqh.test: function to calculate a different variants of the T-squared test of Hotelling (1931)
#' the inputs must be Nx2 or 2xN matrices of numbers, or vectors of complex numbers
#' for a one-sample test, mu is an optional vector to which the data are compared
#' for a two-sample test, x and y are the data from the two conditions
#' if paired=TRUE, the inputs (x and y) must be the same size
#' @export
tsqh.test <- function(x,y=NULL,paired=FALSE,mu=NULL){

  # transpose if dimensions are the wrong way round
  if (!is.null(ncol(x))){d <- dim(x)
  if (d[1]<d[2]){x <- t(x)}}
  if (!is.null(y)){if (!is.null(ncol(y))){d <- dim(y)
  if (d[1]<d[2]){y <- t(y)}}}

  if (!is.null(mu)){for (n in 1:length(mu)){x[,n] <- x[,n] - mu[n]}}

  # convert data to bivariate representation if complex
  if (is.complex(x)){x <- data.frame(Re(x),Im(x))}
  if (!is.null(y)){if (is.complex(y)){y <- data.frame(Re(y),Im(y))}}

  if (is.null(y)){method <- 'One-sample T-squared test'}
  if (!is.null(y) & paired==TRUE){method <- 'Paired T-squared test'}
  if (!is.null(y) & paired==FALSE){method <- 'Independent samples T-squared test'}

  if (paired==TRUE){if (is.null(y)==FALSE){x <- x - y}}

  # paired or one-sample version of the test
  if (paired==TRUE | is.null(y)){

    xbar <- colMeans(x) # estimate means for each DV
    m <- ncol(x)   # record number of DVs
    N <- nrow(x)   # record sample size
    C <- cov(x)    # calculate covariance matrix
    Cinv <- solve(C)  # calculate inverse matrix
    tsq <- N*t(xbar)%*%Cinv%*%xbar # compute t2 statistic
    Fratio <- tsq * (N-m)/(m*(N-1)) # convert to F-ratio
    df1 <- m      # record degrees of freedom
    df2 <- N-m
    pval <- pf(Fratio,df1,df2,lower.tail=FALSE) # estimate p-value

  }

  # independent samples version of the test
  if (paired==FALSE & !is.null(y)){

    xbarA <- colMeans(x) # estimate means for each DV
    m <- ncol(x)   # record number of DVs
    N1 <- nrow(x)   # record sample size

    xbarB <- colMeans(y) # estimate means for each DV
    N2 <- nrow(y)   # record sample size

    CA <- cov(x)    # calculate covariance matrices
    CB <- cov(y)

    meandiff <- xbarA - xbarB

    # combine covariance matrices
    C <- ((N1-1)*CA + (N2-1)*CB)/(N1+N2-m)
    Cinv <- solve(C)                      # calculate inverse matrix

    tsq <- ((N1*N2)/(N1+N2))*t(meandiff)%*%Cinv%*%meandiff    # compute t2 statistic
    Fratio <- tsq * (N1+N2-m-1)/(m*(N1+N2-2))                 # convert to F-ratio
    df1 <- m                                                  # record degrees of freedom
    df2 <- N1+N2-m-1
    pval <- pf(Fratio,df1,df2,lower.tail=FALSE) # estimate p-value

  }

  pval <- min(pval,1)

  # store outputs in a data structure
  output <- data.frame(tsq,Fratio,df1,df2,pval,method)

  return(output)}
